fix: tracking callback invocation

vazarkevych opened the pull request:
## Summary
  
  Fixes duplicate invocation of the tracking callback (`TrackingCallback.onTrack`)
  during **local** experiment evaluation. Previously the callback fired on *every*
  evaluation of the same assignment instead of exactly once.

  ## Problem
  
  `ExperimentEvaluator.isExperimentTracked(...)` recorded each experiment in the
  tracker but **always returned `false`**. Because the call site fires the callback
  when `!isExperimentTracked(...)`, `onTrack` re-fired on every evaluation for the
  same `(hashAttribute, hashValue, experimentKey, variationId)` combination.                                                                                                                           

  ## Fix

  - `isExperimentTracked(...)` now returns whether the assignment was already
    tracked (delegates to the existing `alreadyTracked(...)` helper), so `onTrack`
    fires exactly once per unique assignment.
  - De-duplication is handled solely by the bounded `ExperimentTracker`
    (Guava LRU cache, `maximumSize = 30`).

Commit:
add ExperimentHelper to Options for having ability to distinguish experiment and experiment result for tracking callback invocation

Commit:
Merge branch 'main' into fix-tracking-callback-invocation

# Conflicts:
#	lib/src/main/java/growthbook/sdk/java/evaluators/FeatureEvaluator.java

madhuchavva commented on lib/src/main/java/growthbook/sdk/java/evaluators/ExperimentEvaluator.java:
```diff
@@ -394,23 +391,6 @@ private <ValueType> ExperimentResult<ValueType> getExperimentResult(
-
-        // Add the experiment to the tracker if it doesn't exist.
-        if (!experimentTracker.isExperimentTracked(key)) {
-            experimentTracker.trackExperiment(key);
```
We use an Experiment Tracker to keep track of the most recent experiments set with a bounded max val of 30(which can be changed). Changing this implementation to a unbounded hashSet will grow significantly when there are multiple instances of GrowthBookClient is created.

vazarkevych replied:
We’ve restored your implementation of the Experiment Tracker as suggested.

madhuchavva commented on lib/src/main/java/growthbook/sdk/java/multiusermode/configurations/Options.java:
```diff
@@ -157,6 +157,7 @@ public Options(@Nullable Boolean enabled,
     @Nullable
     private Map<String, Object> globalForcedFeatureValues;
 
+    private ExperimentHelper experimentHelper = new ExperimentHelper();
```
`Options` is not a place for initializing experimentHelper. Options technically provides all configurable parameters that influence the behavior of GrowthBook instance.

vazarkevych replied:
Ok, we have moved the Experiment Tracker to the Global Context. We need this instance to be a singleton in order to track experiments and their results properly — only once. Previously, it was being recreated during every evaluation process

Commit:
isExperimentTracked function is in utils class now, add ExperimentTracker to GlobalContext to make it singleton, remove ExperimentHelper from Options class

vazarkevych committed:
Merge branch 'main' into fix-tracking-callback-invocation

Resolve conflicts by taking main's implementation everywhere, since main's
#216 remote-eval refactor superseded most of this branch's work:
- remote-eval tracking dedup now lives in ExperimentEvaluator.fireRemoteEvaluationTracks()
- tracking-key logic is in private trackingKey()/alreadyTracked() helpers
- ExperimentTracker stays per-evaluator (already long-lived per GrowthBook instance)

Preserve only the still-relevant fix: local isExperimentTracked() in
ExperimentEvaluator always returned false, so the tracking callback re-fired on
every evaluation. It now delegates to alreadyTracked() so repeated
(hashAttribute, hashValue, experimentKey, variationId) combinations are
de-duplicated, matching remote-eval behavior.

NOTE: this changes tracking-callback firing behavior for local evaluation and
should be confirmed with a maintainer before release.

vazarkevych committed:
test: guard tracking-callback de-duplication for local evaluation

Add regression tests proving TrackingCallback.onTrack fires exactly once per
unique (hashAttribute, hashValue, experimentKey, variationId) assignment across
repeated run() calls, and once per distinct experiment. These fail against the
previous isExperimentTracked() that always returned false.

vazarkevych committed:
refactor: remove unused ExperimentHelper (unbounded HashSet tracker)

Experiment de-duplication is handled solely by the bounded ExperimentTracker
(Guava LRU cache, max 30), as requested in review. ExperimentHelper used an
unbounded HashSet and was only instantiated as a dead field in GBContext
(its isTracked() was never called). Removing the class and the field fully
drops the unbounded-tracker remnant.

vazarkevych committed:
Merge branch 'main' into fix-tracking-callback-invocation

# Conflicts:
#	lib/src/main/java/growthbook/sdk/java/evaluators/ExperimentEvaluator.java
